### PR TITLE
Add com.yubico.YubiOath

### PR DESCRIPTION
**Pull request by valentindavid**
Upstream:
https://www.yubico.com/products/services-software/download/yubico-authenticator/
https://github.com/Yubico/yubioath-desktop

I have not asked upstream yet if they want to be involved. Maybe @dagheyman is the person to contact.

### CODE REVIEW

**Comment by LeandroStanger** [comment]
bot, build com.yubico.YubiOath

File: com.yubico.YubiOath.appdata.xml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (barthalion) [inline com.yubico.YubiOath.appdata.xml]: As usually, this should be contributed upstream.

File: com.yubico.YubiOath.json [hunk 2]
```diff
@@ -0,0 +1,268 @@
+{
+    "app-id": "com.yubico.YubiOath",
+    "runtime": "org.kde.Platform",
+    "sdk": "org.kde.Sdk",
+    "runtime-version": "5.12",
+    "command": "yubioath-desktop",
+    "cleanup": [
+	"/include/**",
```
Comment (barthalion) [inline com.yubico.YubiOath.json]: Just `/include` and `/lib/pkgcnfig` should do the trick.

File: com.yubico.YubiOath.json [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+		{
+		    "type": "patch",
+		    "path": "yubioath-desktop-path.patch"
+		}
+            ]
+        },
+	{
+	    "name": "metadata",
+	    "buildsystem": "simple",
+	    "build-commands": [
```
Comment (barthalion) [inline com.yubico.YubiOath.json]: It doesn't need to be a separate module, you can include it in `yubioath-desktop` and install in `post-install`.

**Comment by barthalion** [comment]
bot, build com.yubico.YubiOath

File: com.yubico.YubiOath.appdata.xml (hunk 1 above)
Comment (valentindavid) [reply com.yubico.YubiOath.appdata.xml]: Do you mean that we should ask upstream to distribute this file?

**Comment by LeandroStanger** [comment]
bot, build com.yubico.YubiOath

File: com.yubico.YubiOath.json [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+		"rmdir /app/sbin || true"
+	    ],
+	    "cleanup": [
+		"/lib/libpcscspy.so",
+		"/lib/libpcsclite.so"
+	    ],
+	    "sources": [
+		{
+                    "type": "archive",
+                    "url": "https://github.com/LudovicRousseau/PCSC/archive/pcsc-1.8.24.tar.gz",
```
Comment (Erick555) [inline com.yubico.YubiOath.json]: latest is [1.8.25](https://github.com/LudovicRousseau/PCSC/releases/tag/pcsc-1.8.25)

File: com.yubico.YubiOath.json [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    "runtime": "org.kde.Platform",
+    "sdk": "org.kde.Sdk",
+    "runtime-version": "5.12",
+    "command": "yubioath-desktop",
+    "cleanup": [
+	"/include",
+	"/lib/pkgconfig"
+    ],
+    "finish-args": [
+        "--socket=x11",
```
Comment (Erick555) [inline com.yubico.YubiOath.json]: wayland?

File: com.yubico.YubiOath.json [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+		}
+	    ]
+	},
+	{
+	    "name": "pyotherside",
+            "buildsystem": "qmake",
+	    "sources": [
+		{
+                    "type": "archive",
+                    "url": "https://github.com/thp/pyotherside/archive/1.5.4.tar.gz",
```
Comment (Erick555) [inline com.yubico.YubiOath.json]: latest is [1.5.8](https://github.com/thp/pyotherside/releases/tag/1.5.8)

File: com.yubico.YubiOath.json [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+		"--without-alllang",
+		"--with-python3"
+	    ],
+	    "cleanup": [
+		"/**"
+	    ],
+	    "sources": [
+		{
+                    "type": "archive",
+                    "url": "http://download.sourceforge.net/swig/swig-3.0.12.tar.gz",
```
Comment (Erick555) [inline com.yubico.YubiOath.json]: latest is [4.0.0](https://sourceforge.net/projects/swig/files/swig/swig-4.0.0/)

File: com.yubico.YubiOath.json [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+	{
+            "name": "pyscard",
+            "buildsystem": "simple",
+	    "build-commands": [
+		"pip3 install --prefix /app --no-deps ."
+	    ],
+	    "sources": [
+		{
+		    "type": "archive",
+		    "url": "https://github.com/LudovicRousseau/pyscard/archive/release-1.9.7.tar.gz",
```
Comment (Erick555) [inline com.yubico.YubiOath.json]: latest is [1.9.9](https://github.com/LudovicRousseau/pyscard/releases/tag/release-1.9.9)

File: com.yubico.YubiOath.json [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+	{
+            "name": "cffi",
+            "buildsystem": "simple",
+	    "build-commands": [
+		"pip3 install --prefix /app --no-deps ."
+	    ],
+	    "sources": [
+		{
+		    "type": "archive",
+		    "url": "https://files.pythonhosted.org/packages/64/7c/27367b38e6cc3e1f49f193deb761fe75cda9f95da37b67b422e62281fcac/cffi-1.12.2.tar.gz",
```
Comment (Erick555) [inline com.yubico.YubiOath.json]: latest is [1.12.3](https://bitbucket.org/cffi/cffi/get/v1.12.3.tar.bz2)

File: com.yubico.YubiOath.json [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+	{
+            "name": "cryptography",
+            "buildsystem": "simple",
+	    "build-commands": [
+		"python3 setup.py install --prefix=/app"
+	    ],
+	    "sources": [
+		{
+		    "type": "archive",
+		    "url": "https://github.com/pyca/cryptography/archive/2.6.1.tar.gz",
```
Comment (Erick555) [inline com.yubico.YubiOath.json]: latest is [2.7](https://github.com/pyca/cryptography/releases/tag/2.7)

File: com.yubico.YubiOath.json [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+	{
+            "name": "fido2",
+            "buildsystem": "simple",
+	    "build-commands": [
+		"pip3 install --prefix /app --no-deps ."
+	    ],
+	    "sources": [
+		{
+		    "type": "archive",
+		    "url": "https://github.com/Yubico/python-fido2/releases/download/0.5.0/fido2-0.5.0.tar.gz",
```
Comment (Erick555) [inline com.yubico.YubiOath.json]: latest is [0.7.0](https://github.com/Yubico/python-fido2/releases/tag/0.7.0)

File: com.yubico.YubiOath.json [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+	{
+            "name": "yubikey-manager",
+            "buildsystem": "simple",
+	    "build-commands": [
+		"pip3 install --prefix /app --no-deps ."
+	    ],
+	    "sources": [
+		{
+		    "type": "archive",
+		    "url": "https://github.com/Yubico/yubikey-manager/releases/download/yubikey-manager-0.7.1/yubikey-manager-0.7.1.tar.gz",
```
Comment (Erick555) [inline com.yubico.YubiOath.json]: latest is [3.1.0](https://github.com/Yubico/yubikey-manager/releases/tag/yubikey-manager-3.1.0)!

File: com.yubico.YubiOath.appdata.xml [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+      <image type="source" width="590" height="443">https://raw.githubusercontent.com/Yubico/yubioath-desktop/yubioath-desktop-4.3.5/screenshot.png</image>
+    </screenshot>
+  </screenshots>
+
+  <launchable type="desktop-id">com.yubico.YubiOath.desktop</launchable>
+
+  <url type="homepage">https://developers.yubico.com/yubioath-desktop/</url>
+  <content_rating type="oars-1.1"/>
+  <releases>
+    <release version="4.3.5" date="2019-01-02">
```
[the remark this answers is not in the document]
Comment (valentindavid) [reply com.yubico.YubiOath.appdata.xml]: Fixed

File: com.yubico.YubiOath.json (hunk 4 above)
Comment (valentindavid) [reply com.yubico.YubiOath.json]: Fixed

File: com.yubico.YubiOath.json (hunk 5 above)
Comment (valentindavid) [reply com.yubico.YubiOath.json]: Fixed

File: com.yubico.YubiOath.json (hunk 6 above)
Comment (valentindavid) [reply com.yubico.YubiOath.json]: Fixed

File: com.yubico.YubiOath.json (hunk 8 above)
Comment (valentindavid) [reply com.yubico.YubiOath.json]: Fixed

File: com.yubico.YubiOath.json (hunk 9 above)
Comment (valentindavid) [reply com.yubico.YubiOath.json]: Fixed

File: com.yubico.YubiOath.json (hunk 10 above)
Comment (valentindavid) [reply com.yubico.YubiOath.json]: Fixed

File: com.yubico.YubiOath.json (hunk 11 above)
Comment (valentindavid) [reply com.yubico.YubiOath.json]: Fixed

File: com.yubico.YubiOath.json (hunk 12 above)
Comment (valentindavid) [reply com.yubico.YubiOath.json]: Fixed

**Comment by valentindavid** [comment]
bot, build com.yubico.YubiOath

File: com.yubico.YubiOath.appdata.xml (hunk 1 above)
Comment (bilelmoussaoui) [reply com.yubico.YubiOath.appdata.xml]: Yes. It's a freedesktop standard & not only a Flathub thing. Other packages could benefit from a nice appdata file.

File: com.yubico.YubiOath.json [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+		    "url": "https://github.com/Yubico/yubikey-manager/archive/yubikey-manager-3.1.0.tar.gz",
+		    "sha256": "9009ad6695b262c23abfb2d030786bcb5d72756f822ce74eb20d611a9cd130bd"
+		}
+	    ]
+	},
+        {
+            "name": "yubioath-desktop",
+            "buildsystem": "qmake",
+	    "post-install": [
+		"sed -i 's/Icon=yubioath/Icon=com.yubico.YubiOath/' resources/yubioath-desktop.desktop",
```
Comment (bilelmoussaoui) [inline com.yubico.YubiOath.json]: This should be proposed upstream too. It doesn't make sense for an app to keep two different app-ids :)

File: com.yubico.YubiOath.json [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+		"install -Dm644 resources/yubioath-desktop.desktop /app/share/applications/com.yubico.YubiOath.desktop",
+		"install -Dm644 resources/icons/yubioath.png /app/share/icons/hicolor/128x128/apps/com.yubico.YubiOath.png",
+                "install -Dm644 com.yubico.YubiOath.appdata.xml /app/share/appdata/com.yubico.YubiOath.appdata.xml"
```
Comment (bilelmoussaoui) [inline com.yubico.YubiOath.json]:
```suggestion
                "install -Dm644 com.yubico.YubiOath.appdata.xml /app/share/metainfo/com.yubico.YubiOath.appdata.xml"
```
Metainfo is the new place where the appdata files should be installed.

Comment (Erick555) [reply com.yubico.YubiOath.json]: @bilelmoussaoui is it documented somewhere?

Comment (bilelmoussaoui) [reply com.yubico.YubiOath.json]: Yes, the specs are here https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html

**Comment by bilelmoussaoui** [comment]
> Considering there is a shitload of deps, maintaining all of them will be quite challenging without upstream involvement.
@Erick555 The python dependencies could be automated using flatpak-builder-tools

**Comment by bilelmoussaoui** [comment]
Have you got the time to contact upstream regarding this MR?

**Comment by dagheyman** [comment]
What would be required/requested from upstream in this scenario? I don't have any experience with the flatpak format currently.

File: com.yubico.YubiOath.json [hunk 16]
```diff
@@ -0,0 +1,258 @@
+{
+    "app-id": "com.yubico.YubiOath",
+    "runtime": "org.kde.Platform",
+    "sdk": "org.kde.Sdk",
+    "runtime-version": "5.12",
```
Comment (LeandroStanger) [inline com.yubico.YubiOath.json]: Update runtime to 5.13

Comment (nedrichards) [reply com.yubico.YubiOath.json]: That is not necessarily required. It'd be good to check.

**Comment by nedrichards** [comment]
@dagheyman basically 'keep an eye on it' and be as involved if they wish. But in this case nothing would be required, we'd just like to make sure upstream has access in order to make sure that the app uses the tested set of dependencies and makes them happy etc. (or can complain if that doesn't happen).

**Comment by dagheyman** [comment]
@nedrichards Sounds good, feel free to open issues in yubico/yubioath-desktop if there is anything more specific.

**Comment by bilelmoussaoui** [comment]
@dagheyman We would love to see the appdata file upstreamed first & probably the other patches that arelative to yubioath-desktop. Any of you upstream wants write access to the repository once created?

**Comment by bilelmoussaoui** [comment]
com.yubico.YubiOath

**Comment by bilelmoussaoui** [comment]
bot, build com.yubico.YubiOath

**Comment by dagheyman** [comment]
We're open to upstreaming the appdata file (PRs welcome), but would prefer not to have commit access since that may give the impression we're committing to supporting this packaging. We'll treat the flatpak as "not supported or maintained by us, but happy to help out if needed", just like we treat the arch packages, fedora packages, debian packages etc.

File: com.yubico.YubiOath.appdata.xml [hunk 17]
```diff
@@ -0,0 +1,46 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<application>
```
Comment (bilelmoussaoui) [inline com.yubico.YubiOath.appdata.xml]: I would run `appstream-util upgrade` or just replace this with `​<component type="desktop-application">`

**Comment by ai** [comment]
How I can help with moving it to Flathub?

**Comment by bilelmoussaoui** [comment]
> How I can help with moving it to Flathub?

You can re-submit this PR based on @valentindavid 's work and fixing the rest of the requested changes. Upstream doesn't seem to want write access to the flatpak repository, but they are okay with someone from the community doing that per https://github.com/Yubico/yubioath-desktop/issues/506.

**Comment by Erick555** [comment]
Note that without upstream involved there is no guarantee someone won't lose interest in a few months and leave flatpak for rot which isn't uncommon and one of the reasons why upstream involvement is so valuable.

**Comment by ai** [comment]
I will try to update PR after Jan 15 (and backport the most useful changes back to Yubico).

Do you know a great guide about Flatpak packaging to read before?

File: com.yubico.YubiOath.json [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+                    "url": "https://developers.yubico.com/yubioath-desktop/Releases/yubioath-desktop-4.3.6.tar.gz",
+		    "sha256": "451d47f23d08dbe3630dcaa14bf11bbd8e894c86e9dee6264af597b5a4550468"
+                },
+		{
+		    "type": "patch",
+		    "path": "yubioath-desktop-path.patch"
+		},
+                {
+		    "type": "file",
+		    "path": "com.yubico.YubiOath.appdata.xml"
```
Comment (LeandroStanger) [inline com.yubico.YubiOath.json]: New appdata files should use metainfo instead. https://www.freedesktop.org/software/appstream/docs/chap-Quickstart.html

File: com.yubico.YubiOath.json (hunk 14 above)
Comment (ai) [reply com.yubico.YubiOath.json]: I sent PR to upstream https://github.com/Yubico/yubioath-desktop/pull/519

Comment (ai) [reply com.yubico.YubiOath.json]: PR was accepted

File: com.yubico.YubiOath.appdata.xml (hunk 1 above)
Comment (ai) [reply com.yubico.YubiOath.appdata.xml]: I sent PR to upstream with fixes suggested below https://github.com/Yubico/yubioath-desktop/pull/524

**Comment by bilelmoussaoui** [comment]
This was merged in #1451

**Pull request closed without merging**